### PR TITLE
Run by Line: disconnect at last line

### PR DESCRIPTION
**Pull request by DavidKutu**
For #6858

-   [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR).
-   [x] Title summarizes what is changing.
-   [x] Has a [news entry](https://github.com/Microsoft/vscode-jupyter/tree/main/news) file (remember to thank yourself!).
-   [x] Appropriate comments and documentation strings in the code.
-   [ ] Has sufficient logging.
-   [ ] Has telemetry for enhancements.
-   [ ] Unit tests & system/integration tests are added/updated.
-   [ ] [Test plan](https://github.com/Microsoft/vscode-jupyter/blob/main/.github/test_plan.md) is updated as appropriate.
-   [ ] [`package-lock.json`](https://github.com/Microsoft/vscode-jupyter/blob/main/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed).

### CODE REVIEW

File: src/client/debugger/jupyter/kernelDebugAdapter.ts [hunk 1]
```diff
@@ -313,6 +335,25 @@ export class KernelDebugAdapter implements DebugAdapter, IKernelDebugAdapter {
             }
         });
 
+        if ((message as DebugProtocol.StackTraceResponse).command === 'stackTrace') {
+            (message as DebugProtocol.StackTraceResponse).body.stackFrames.forEach((sf) => {
+                // Check if we're stopped at the last line
+                let currentCell: NotebookCell | undefined;
+                this.notebookDocument.getCells().forEach((cell) => {
+                    const index = sf.source?.path!.indexOf('#ch');
```
Comment (DonJayamanne) [inline src/client/debugger/jupyter/kernelDebugAdapter.ts:343]: this should be a helper function, give a stack we should find the corresponding cell.
I'm assuming we have similar code today elsewhere.
Also no idea what `#ch` is here.

Comment (DonJayamanne) [inline src/client/debugger/jupyter/kernelDebugAdapter.ts:343]: This code can go outside the forEach loop, this doesn't change, hence no need to get index & fragment everytime for each cell.

File: src/client/debugger/jupyter/kernelDebugAdapter.ts [hunk 2]
```diff
@@ -116,8 +117,11 @@ export class KernelDebugAdapter implements DebugAdapter, IKernelDebugAdapter {
             // eslint-disable-next-line @typescript-eslint/no-explicit-any
             const content = msg.content as any;
             if (content.event === 'stopped') {
-                this.runByLineThreadId = content.body.threadId;
-                this.runByLineSeq = content.seq;
+                if (this.isRunByLine) {
```
Comment (DonJayamanne) [inline src/client/debugger/jupyter/kernelDebugAdapter.ts:120]: Does this code work if i have a cell as follows:
```
def do_something():
    print("Hello")


do_something()
```

Now, when I'm at the last line, will I be able to step in?

Comment (DonJayamanne) [reply src/client/debugger/jupyter/kernelDebugAdapter.ts:120]: Or what if i add a breakpoint at line `print("Hello")`
& then hit `continue`, surely it shouldn't stop the run by line, but should step into that code.

File: src/client/debugger/jupyter/kernelDebugAdapter.ts (hunk 1 above)
Comment (DavidKutu) [reply src/client/debugger/jupyter/kernelDebugAdapter.ts:343]: I'll put it on a helper. The `#ch` is to find the fragment in the stackFrame

File: src/client/debugger/jupyter/kernelDebugAdapter.ts (hunk 2 above)
Comment (DavidKutu) [reply src/client/debugger/jupyter/kernelDebugAdapter.ts:120]: No, I don't remember how it used to work, I'll check

Comment (DavidKutu) [reply src/client/debugger/jupyter/kernelDebugAdapter.ts:120]: Looks like run by line in the webview doesn't work on stable. I'll go for it anyway.

Comment (DavidKutu) [reply src/client/debugger/jupyter/kernelDebugAdapter.ts:120]: The new approach covers that scenario

**Review by rchiodo — APPROVED** [review]

**Pull request merged**